fix(compile): do not create directives for constant media URL attributes

petebacondarwin opened the pull request:
By creating attribute directives that watch the value of
media url attributes (e.g. `img[src]`) we caused a conflict
when both `src` and `data-src` were appearing on the
same element. As each directive was trying to write to the
attributes on the element, where AngularJS treats `src` and
`data-src` as synonymous.

This commit ensures that we do not create create such directives
when the media url attribute is a constant (no interpolation).

Because of this (and because we no longer sanitize URLs in the
`$attr.$set()` method, this commit also updates `ngHref` and
`ngSrc` to do a preliminary sanitization of URLs in case there
is no interpolation in the attribute value.

jbedard commented on src/ng/compile.js:
```diff
@@ -3871,6 +3871,12 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
               pre: function attrInterpolatePreLinkFn(scope, element, attr) {
+                if (trustedContext === $sce.MEDIA_URL && isConstantInterpolation(interpolateFn)) {
+                  // Constant MEDIA_URL attributes should just be sanitized and not watched.
+                  attr[name] = interpolateFn(scope);
```
Do we still need this assignment? I thought to go back to the previous behavior we'd want to do *no* assignments.

I thought we'd just update [this condition](https://github.com/angular/angular.js/blob/76f3bb0e2bc5a781f06546485fdb9dc46ef9a2de/src/ng/interpolate.js#L245) to return early like it previously did...

petebacondarwin replied:
This was the only way to keep all the tests passing. If you change the condition you link to the a load of other ngSrc etc related tests fail. This is due to how url sanitation has moved to the interpolation rather than being done surely in the compiler.

jbedard replied:
Oh :/

And we can't do this `if + return` outside the directive? Before doing `directives.push`...

fbrandel commented:
When I run the fix locally, it seems that the `ng-attr-data-src` directive now stopped working. The attribute is not interpolated/watched, but it should. Unfortunately I was not able to provide a test case for this, just noticed it by manually trying it out. 

The original reported bug was fixed though :)

petebacondarwin commented:
@fbrandel - I don't think what you are suggesting ever worked with Angular. The following variants on `ng-attr` work: `ng-attr-src` and `data-ng-attr-src` but `ng-attr-data-src` does not work. See http://next.plnkr.co/edit/sbrIYJYyKE0S7XVA?preview

petebacondarwin commented:
Ahah! Let me see if I can reproduce in a unit test...

petebacondarwin commented:
@fbrandel I added unit tests for this case but I cannot reproduce your issue. Please can you take a look and see if you can do so?

fbrandel commented:
@petebacondarwin After building the project with the PR applied, I can see the tests pass.

But there is one strange thing: When I use the _minified_ angular.js file on my local test page, its works, but not with the _un-minified_ version. Not sure if this is a different/local problem and therefore has nothing to to do with the bug. Unfortunately I was not able to setup a Plunkr as the un-minified version is somehow too large to be delivered.

fbrandel commented:
@petebacondarwin Sorry to come up with more investigation ;) It seems that the combination of `src` and `ng-attr-data-src` attributes on an image now behaves differently. With a `src` attribute set on the image, the `data-src` does not contain the value of the `ng-attr-data-src` expression but the value of `src`.

Plunkr: https://next.plnkr.co/edit/mMn0sv47O81Hi0FJ

Unit Test:
```js
it('should compile img with constant [src]-attribute and [ng-attr-data-src] attribute', inject(function() {
        $rootScope.name = 'some-image.png';
        element = $compile('<img src="constant.png" ng-attr-data-src="{{name}}">')($rootScope);
        expect(element.attr('data-src')).toBeUndefined();

        $rootScope.$digest();
        expect(element.attr('src')).toBe('constant.png');
        expect(element.attr('data-src')).toBe('some-image.png');

        $rootScope.name = 'other-image.png';
        $rootScope.$digest();
        expect(element.attr('src')).toBe('constant.png');
        expect(element.attr('data-src')).toBe('other-image.png');
}));
```

petebacondarwin commented:
Hmm, oh dear. This is a can of worms...

petebacondarwin committed:
fix(interpolate): do not create directives for constant media URL attributes

By creating attribute directives that watch the value of
media url attributes (e.g. `img[src]`) we caused a conflict
when both `src` and `data-src` were appearing on the
same element. As each directive was trying to write to the
attributes on the element, where AngularJS treats `src` and
`data-src` as synonymous.

This commit ensures that we do not create create such directives
when the media url attribute is a constant (no interpolation).

Because of this (and because we no longer sanitize URLs in the
`$attr.$set()` method, this commit also updates `ngHref` and
`ngSrc` to do a preliminary sanitization of URLs in case there
is no interpolation in the attribute value.

Fixes #16734

petebacondarwin commented:
@fbrandel - can you take a look at this alternative approach? See if you can break this one :-)

fbrandel commented:
@petebacondarwin Looking good with my scenarios! Thank you so far for taking care!

mgol commented on test/ng/compileSpec.js:
```diff
@@ -12109,6 +12118,47 @@ describe('$compile', function() {
         expect(element.attr('dash-test4')).toBe('JamieMason');
       }));
 
+      it('should work with img[src]', inject(function() {
```
It tests ng-attr-src, not src. The same remark for the next test.

petebacondarwin committed:
fixup! fix(interpolate): do not create directives for constant media URL attributes

Narretz commented:
We are now setting the attribute twice, even for interpolated attributes, which is not that great imo. Can we detect if the attribute value has not been interpolated?

gkalpak commented on src/ng/directive/attrs.js:
```diff
@@ -422,6 +422,10 @@ forEach(['src', 'srcset', 'href'], function(attrName) {
 
+        // We need to sanitize the url at least once, in case it is a constant
+        // non-interpolated attribute.
+        attr.$set(normalized, $sce.getTrustedMediaUrl(attr[normalized]));
```
Maybe we could check the current value and only call `$set()` if `getTrustedMediaUrl()` return a different value.

petebacondarwin committed:
fixup! fix(interpolate): do not create directives for constant media URL attributes
